src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(stateCopy, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach(function (key) {
+        delete stateCopy[key];
+      });
+    }
+
+    if (action.type === 'clear') {
+      for (const key in stateCopy) {
+        if (stateCopy.hasOwnProperty(key)) {
+          delete stateCopy[key];
+        }
+      }
+    }
+
+    currentState.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF]: It is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This is already done correctly here, so no changes are needed for this part.